 Завантаження бібліотеки Pandas
У цьому кроці ми встановлюємо та імпортуємо бібліотеку pandas,
яка використовується для роботи з таблицями в Python.

In [1]:
import pandas as pd

In [3]:
pd.__version__

'2.2.3'

Завантажуємо датасети
В зв'язку з тим що датасети заархівовані, проводимо розархівування 

In [5]:
import zipfile
import os

zip_name = "stack-overflow-developer-survey-2025.zip"

with zipfile.ZipFile(zip_name, 'r') as zip_ref:
    zip_ref.extractall("so_survey")

print("Розпаковано в папку so_survey")
os.listdir("so_survey")

Розпаковано в папку so_survey


['2025_Developer_Survey_Tool.pdf',
 'survey_results_public.csv',
 'survey_results_schema.csv']

In [2]:
public = pd.read_csv("so_survey/survey_results_public.csv")
schema = pd.read_csv("so_survey/survey_results_schema.csv")

C:\Users\ПК\AppData\Local\Temp\ipykernel_10192\1863137603.py:1: DtypeWarning: Columns (56,74,92,97,98,105,109,110,132,162,165) have mixed types. Specify dtype option on import or set low_memory=False.
  public = pd.read_csv("so_survey/survey_results_public.csv")


Перевіряємо заватаження датасетів. Перші рядки.

In [7]:
print("Public shape:", public.shape)
print("Schema shape:", schema.shape)

Public shape: (49191, 172)
Schema shape: (139, 6)


In [8]:
public.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


In [9]:
schema.head()

,qid,qname,question,type,sub,sq_id
0,QID18,TechEndorse_1,What attracts you to a technology or causes yo...,RO,AI integration or AI Agent capabilities,1.0
1,QID18,TechEndorse_2,What attracts you to a technology or causes yo...,RO,Easy-to-use API,2.0
2,QID18,TechEndorse_3,What attracts you to a technology or causes yo...,RO,Robust and complete API,3.0
3,QID18,TechEndorse_4,What attracts you to a technology or causes yo...,RO,Customizable and manageable codebase,4.0
4,QID18,TechEndorse_5,What attracts you to a technology or causes yo...,RO,Reputation for quality,5.0


 Назви колонок. 

In [10]:
print(public.columns[:10])
print("Кількість колонок:", len(public.columns))

Index(['ResponseId', 'MainBranch', 'Age', 'EdLevel', 'Employment',
       'EmploymentAddl', 'WorkExp', 'LearnCodeChoose', 'LearnCode',
       'LearnCodeAI'],
      dtype='object')
Кількість колонок: 172


Типи даних.

In [11]:
public.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49191 entries, 0 to 49190
Columns: 172 entries, ResponseId to JobSat
dtypes: float64(52), int64(1), object(119)
memory usage: 64.6+ MB


Швидка перевірка пропусків

In [12]:
public.isnull().sum().sort_values(ascending=False).head(10)

AIAgentObsWrite         48927
SOTagsWant Entry        48761
SOTagsHaveEntry         48733
AIModelsWantEntry       48716
AIAgentOrchWrite        48713
JobSatPoints_15_TEXT    48527
AIAgentKnowWrite        48425
AIModelsHaveEntry       48415
SO_Actions_15_TEXT      48368
AIAgentExtWrite         48332
dtype: int64

Завдання 1. Підрахунок загальної кількості респондентів які приймали участь в опитуванні Stack Overflow.

In [13]:
total_respondents = public.shape[0]

print("Загальна кількість респондентів:", total_respondents)

Загальна кількість респондентів: 49191


Завдання 2. Аналіз повноти відповідей респондентів
Мета: порахувати кількість респондентів, у яких немає пропусків у всіх питаннях опитування.

1) Отримуємо список питань з survey_results_schema (поле qname)


In [15]:
qnames = schema["qname"].dropna().unique()
print("Кількість qname у schema:", len(qnames))
print("Перші 10 qname:", qnames[:10])

Кількість qname у schema: 139
Перші 10 qname: ['TechEndorse_1' 'TechEndorse_2' 'TechEndorse_3' 'TechEndorse_4'
 'TechEndorse_5' 'TechEndorse_6' 'TechEndorse_7' 'TechEndorse_8'
 'TechEndorse_9' 'TechEndorse_13']


2. Інтерсекція множин (фільтруємо колонки). Беремо тільки ті qname, які реально є в public.

In [16]:
public_cols = set(public.columns)
schema_cols = set(qnames)

common_cols = sorted(public_cols.intersection(schema_cols))

print("Кількість колонок у public:", len(public.columns))
print("Кількість qname у schema:", len(qnames))
print("Кількість спільних колонок (public ∩ schema):", len(common_cols))
print("Перші 10 спільних колонок:", common_cols[:10])

Кількість колонок у public: 172
Кількість qname у schema: 139
Кількість спільних колонок (public ∩ schema): 126
Перші 10 спільних колонок: ['AIAcc', 'AIAgentChange', 'AIAgentExtWrite', 'AIAgentExternal', 'AIAgentKnowWrite', 'AIAgentKnowledge', 'AIAgentObsWrite', 'AIAgentObserveSecure', 'AIAgentOrchWrite', 'AIAgentOrchestration']


3. Підрахунок респондентів без пропусків

In [17]:
# Беремо тільки потрібні колонки
public_common = public[common_cols]

# Маска: True, якщо в рядку немає пропусків у жодній з цих колонок
no_missing_mask = public_common.notna().all(axis=1)

# Підрахунок респондентів, які відповіли на всі запитання
respondents_all_answered = int(no_missing_mask.sum())

print(respondents_all_answered)

0


Перевірка 1: скільки колонок реально перевіряємо

In [18]:
print("К-сть qname у schema:", schema["qname"].nunique())
print("К-сть колонок у public:", public.shape[1])
print("К-сть спільних колонок (public ∩ schema):", len(common_cols))

К-сть qname у schema: 139
К-сть колонок у public: 172
К-сть спільних колонок (public ∩ schema): 126


Перевірка 2: максимум скільки питань взагалі заповнив "найповніший" респондент

In [19]:
filled_per_row = public[common_cols].notna().sum(axis=1)

print("Максимум заповнених відповідей у одного респондента:", filled_per_row.max())
print("К-сть питань, які ми перевіряємо:", len(common_cols))

Максимум заповнених відповідей у одного респондента: 122
К-сть питань, які ми перевіряємо: 126


Так як filled_per_row.max() менше , ніж len(common_cols) - тоді 0 гарантовано і це підтвердження

Перевірка 3: Які колонки найчастіше "ламають"повну відповідь

In [20]:
missing_by_col = public[common_cols].isna().sum().sort_values(ascending=False)

print("Топ-10 колонок з найбільшою кількістю пропусків:")
print(missing_by_col.head(10))

Топ-10 колонок з найбільшою кількістю пропусків:
AIAgentObsWrite         48927
SOTagsWant Entry        48761
SOTagsHaveEntry         48733
AIModelsWantEntry       48716
AIAgentOrchWrite        48713
JobSatPoints_15_TEXT    48527
AIAgentKnowWrite        48425
AIModelsHaveEntry       48415
SO_Actions_15_TEXT      48368
AIAgentExtWrite         48332
dtype: int64


Отже

In [21]:
respondents_all_answered = int(public[common_cols].notna().all(axis=1).sum())
print(respondents_all_answered)

0


Завдання 3. Статистичний аналіз досвіду респондентів.
Мета: обчислити міри центральної тенденції для змінної WorkExp
(досвід роботи респондентів у роках).
Показники:
- середнє (mean)
- медіана (median)
- мода (mode)


In [4]:
#Перевіряємо що колонка існує
print("WorkExp" in public.columns)
# Дивимось тип і кількість пропусків
public["WorkExp"].info()


True
<class 'pandas.core.series.Series'>
RangeIndex: 49191 entries, 0 to 49190
Series name: WorkExp
Non-Null Count  Dtype  
--------------  -----  
42893 non-null  float64
dtypes: float64(1)
memory usage: 384.4 KB


Очищення. Ігноруємо пропуски.

In [5]:
work_exp = public["WorkExp"].dropna()

print("Кількість значень без пропусків:", work_exp.shape[0])
# Пропущені значення були виключені з аналізу, оскільки вони не містять числової інформації і можуть спотворювати результати обчислення мір центральної тенденції.

Кількість значень без пропусків: 42893


Обчислення статистик

In [6]:
mean_exp = work_exp.mean()
median_exp = work_exp.median()
mode_exp = work_exp.mode().iloc[0]  # беремо першу моду, якщо їх кілька

stats = pd.DataFrame({
    "Показник": ["Середнє", "Медіана", "Мода"],
    "Значення": [mean_exp, median_exp, mode_exp]
})

stats

,Показник,Значення
0,Середнє,13.367403
1,Медіана,10.000000
2,Мода,10.000000


Виснновок.
Середній досвід роботи респондентів становить 13.37 року, тоді як медіана і мода дорівнюють 10 рокам.
Це означає, що більшість учасників опитування мають близько 10 років досвіду, а вищий середній показник зумовлений наявністю респондентів із дуже великим стажем, які “тягнуть” середнє значення вгору.
Отже, розподіл досвіду є асиметричним вправо, з невеликою кількістю дуже досвідчених спеціалістів.


Завдання 4. Аналіз віддаленої роботи.
Мета: визначити кількість респондентів, які працюють віддалено (remote).
Для цього використовується відповідна колонка з датасету public.

Знайдемо колонку з форматом роботи

In [7]:
print("RemoteWork" in public.columns)

True


Подивимось які там значення

In [9]:
public["RemoteWork"].value_counts(dropna=False)

RemoteWork
NaN                                                                             15411
Remote                                                                          10931
Hybrid (some remote, leans heavy to in-person)                                   6732
In-person                                                                        6042
Hybrid (some in-person, leans heavy to flexibility)                              5831
Your choice (very flexible, you can come in when you want or just as needed)     4244
Name: count, dtype: int64

Фільтруємо лише повністю віддалених

In [11]:
remote_only = public[public["RemoteWork"] == "Remote"]
remote_count = remote_only.shape[0]

print(remote_count)

10931


Висновок.
10 931 респондент працює повністю віддалено.

Завдання 5. Визначення популярності Python
Мета: визначити відсоток респондентів, які програмують на Python.
Оскільки в полі LanguageHaveWorkedWith може бути кілька мов,
потрібно перевірити наявність "Python" у рядку.

In [3]:
# Перевірка, що колонка існує
print("LanguageHaveWorkedWith" in public.columns)

True


In [5]:
# Загальна кількість респондентів (без пропусків у цій колонці)
total_with_langs = public["LanguageHaveWorkedWith"].notna().sum()

In [6]:
# Маска: чи є Python у списку мов
python_mask = public["LanguageHaveWorkedWith"].str.contains("Python", na=False)

In [7]:
# Кількість респондентів, що використовують Python
python_users = python_mask.sum()

In [3]:
# Відсоток
total_respondents = public.shape[0] 
python_users = public["LanguageHaveWorkedWith"].str.contains("Python", na=False).sum() 
python_percent = (python_users / total_respondents) * 100 
print(f"{python_percent:.1f}%") 

37.5%


Висновок
Близько 37,5% респондентів зазначили, що використовують Python у своїй роботі або навчанні. Це означає, що понад кожен третій респондент працює з Python, що підтверджує його високу популярність та універсальність у сучасній ІТ-індустрії.

Завдання 6. Аналіз шляхів навчання програмуванню
Мета: визначити кількість респондентів, які навчалися програмувати за допомогою онлайн курсів.
Для цього використовується колонка LearnCode, яка може містити множинні значення.


In [9]:
# Перевірка, що колонка існує
print("LearnCode" in public.columns)

True


In [10]:
# Дивимось які значення в цій колонці
public["LearnCode"].head()

0    Online Courses or Certification (includes all ...
1    Online Courses or Certification (includes all ...
2    Online Courses or Certification (includes all ...
3    Other online resources (e.g. standard search, ...
4                                                  NaN
Name: LearnCode, dtype: object

In [4]:
# Маска: чи є "Online" у способах навчання
online_courses_mask = public["LearnCode"].str.contains("Online Courses", na=False) 
online_courses_count = online_courses_mask.sum() 
print(online_courses_count)

10973


Висновок
10 973 респонденти навчалися програмувати за допомогою онлайн курсів.
Це свідчить про те, що структуровані онлайн програми навчання відіграють важливу роль у підготовці спеціалістів у сфері ІТ, навіть на тлі великої кількості альтернативних шляхів навчання.

Завдання 7. Географічний аналіз компенсації Python-розробників
Мета: серед респондентів, які програмують на Python, проаналізувати річну компенсацію
(ConvertedCompYearly) в розрізі країн. Для кожної країни обчислюються
середнє та медіанне значення компенсації.

In [12]:
# Перевірка потрібних колонок
cols = ["LanguageHaveWorkedWith", "ConvertedCompYearly", "Country"]
print(all(col in public.columns for col in cols))

public[cols].head()

True


,LanguageHaveWorkedWith,ConvertedCompYearly,Country
0,Bash/Shell (all shells);Dart;SQL,61256.0,Ukraine
1,Java,104413.0,Netherlands
2,Dart;HTML/CSS;JavaScript;TypeScript,53061.0,Ukraine
3,Java;Kotlin;SQL,36197.0,Ukraine
4,C;C#;C++;Delphi;HTML/CSS;Java;JavaScript;Lua;P...,60000.0,Ukraine


In [13]:
# Фільтр Python + без пропусків у зарплаті
python_df = public[
    public["LanguageHaveWorkedWith"].str.contains("Python", na=False)
]

python_df = python_df.dropna(subset=["ConvertedCompYearly", "Country"])

print("Кількість Python-розробників з відомою компенсацією:", python_df.shape[0])

Кількість Python-розробників з відомою компенсацією: 12765


In [14]:
# Групування по країнах
country_stats = (
    python_df
    .groupby("Country")["ConvertedCompYearly"]
    .agg(["mean", "median"])
    .reset_index()
)

country_stats.columns = ["Country", "MeanCompensation", "MedianCompensation"]

country_stats.head()

,Country,MeanCompensation,MedianCompensation
0,Afghanistan,22328.666667,1000.0
1,Albania,47217.600000,50000.0
2,Algeria,20187.285714,7088.0
3,Andorra,226103.500000,226103.5
4,Antigua and Barbuda,1.000000,1.0


In [15]:
# Відсортувати за середньою компенсацією
country_stats.sort_values("MeanCompensation", ascending=False).head(10)

,Country,MeanCompensation,MedianCompensation
102,Oman,390135.000000,390135.0
3,Andorra,226103.500000,226103.5
149,Viet Nam,218837.166667,8254.0
145,United States of America,173298.590211,150000.0
132,Switzerland,156456.600000,142592.0
121,Singapore,147515.121951,91391.0
65,Israel,135828.365385,142594.0
98,Nigeria,128773.276596,1808.0
63,Ireland,120523.918919,116015.0
99,Nomadic,120131.571429,139218.0


Висновок
Серед Python-розробників найбільші середні та медіанні рівні компенсації спостерігаються в країнах з більш прогресивними ІТ-ринками та високою вартістю життя.
У топі за середньою компенсацією знаходяться Oman, Andorra, Viet Nam, United States та Switzerland.
Водночас різниця між середнім та медіанним значеннями у багатьох країнах (наприклад, United States, Singapore, Israel) свідчить про нерівномірний розподіл доходів: невелика кількість дуже високооплачуваних спеціалістів значно підвищує середнє значення.
Отже, медіана є більш підходящим показником типового доходу Python-розробника, ніж середнє.

Завдання 8. Аналіз освіти найбільш оплачуваних спеціалістів
Мета: визначити рівні освіти 5 респондентів з найбільшою річною компенсацією
(ConvertedCompYearly).

In [16]:
# Перевірка потрібних колонок
cols = ["ConvertedCompYearly", "EdLevel"]
print(all(col in public.columns for col in cols))

public[cols].head()

True


,ConvertedCompYearly,EdLevel
0,61256.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"
1,104413.0,"Associate degree (A.A., A.S., etc.)"
2,53061.0,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)"
3,36197.0,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)"
4,60000.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"


In [17]:
# Прибираємо пропуски
comp_edu = public.dropna(subset=["ConvertedCompYearly", "EdLevel"])
print("Рядків без пропусків:", comp_edu.shape[0])

Рядків без пропусків: 23930


In [18]:
# Сортуємо та обираємо топ 5
top5 = comp_edu.sort_values("ConvertedCompYearly", ascending=False).head(5)

top5[["ConvertedCompYearly", "EdLevel"]]

,ConvertedCompYearly,EdLevel
34267,50000000.0,"Associate degree (A.A., A.S., etc.)"
28700,33552715.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"
43143,18387548.0,"Associate degree (A.A., A.S., etc.)"
35353,15430267.0,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)"
45971,13921760.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"


Висновок
Серед 5 респондентів з найбільшою річною компенсацією переважають такі рівні освіти:
Master’s degree (M.A., M.S., M.Eng., MBA, etc) — 2 особи
Associate degree (A.A., A.S., etc) — 2 особи
Bachelor’s degree (B.A., B.S., B.Eng., etc) — 1 особа
Це свідчить про те, що високі доходи можливі не лише з магістерською освітою, але й з іншими рівнями, включаючи асоційований та бакалаврський ступінь. Формальна освіта є важливою, але не є єдиним фактором, що визначає рівень компенсації в ІТ.

Завдання 9. Аналіз популярності Python по віковим категоріям 
Мета: визначити, який відсоток респондентів у кожній віковій категорії
програмує на Python.

In [19]:
# Перевірка колонок
cols = ["Age", "LanguageHaveWorkedWith"]
print(all(col in public.columns for col in cols))

public[cols].head()

True


,Age,LanguageHaveWorkedWith
0,25-34 years old,Bash/Shell (all shells);Dart;SQL
1,25-34 years old,Java
2,35-44 years old,Dart;HTML/CSS;JavaScript;TypeScript
3,35-44 years old,Java;Kotlin;SQL
4,35-44 years old,C;C#;C++;Delphi;HTML/CSS;Java;JavaScript;Lua;P...


In [20]:
# Створюємо маску Python
python_mask = public["LanguageHaveWorkedWith"].str.contains("Python", na=False)

In [21]:
#Групування по вікових категоріях
age_python = (
    public
    .assign(PythonUser=python_mask)
    .groupby("Age")["PythonUser"]
    .mean()
    .reset_index()
)

age_python["PythonPercent"] = age_python["PythonUser"] * 100

age_python = age_python[["Age", "PythonPercent"]]

age_python

,Age,PythonPercent
0,18-24 years old,40.000000
1,25-34 years old,36.939282
2,35-44 years old,36.719281
3,45-54 years old,38.629482
4,55-64 years old,37.242955
5,65 years or older,31.634820
6,Prefer not to say,31.216931


Висновок
Найвищий відсоток Python-розробників спостерігається у віковій групі 18–24 роки (40%).
Далі показники поступово знижуються з віком.
Найменша частка Python-користувачів — серед респондентів 65+ років (~31.6%).
Отже, Python є найпопулярнішим серед молодших спеціалістів, але зберігає високу частку і в старших вікових групах, що свідчить про його універсальність та актуальність на всіх етапах кар’єри.

Завдання 10. Аналіз індустрій серед високооплачуваних віддалених працівників
Мета: визначити, в яких індустріях найчастіше працюють респонденти,
які входять до 75-го перцентиля за компенсацією та працюють віддалено.

In [22]:
#75-перцентль компенсації. Беремо тільки не порожні компенсації
comp = public["ConvertedCompYearly"].dropna()

p75 = comp.quantile(0.75)

print("75 перцентиль компенсації:", p75)

75 перцентиль компенсації: 120596.0


In [23]:
# Фільтр: висока компенсація + remote
high_paid_remote = public[
    (public["ConvertedCompYearly"] >= p75) &
    (public["RemoteWork"] == "Remote")
]

print("Кількість високооплачуваних remote:", high_paid_remote.shape[0])

Кількість високооплачуваних remote: 2461


In [24]:
# Визначаємо індустрії.У Stack Overflow колонка Industry
print("Industry" in public.columns)

high_paid_remote = high_paid_remote.dropna(subset=["Industry"])

industry_counts = (
    high_paid_remote["Industry"]
    .value_counts()
    .reset_index()
)

industry_counts.columns = ["Industry", "Count"]

industry_counts

True


,Industry,Count
0,Software Development,1186
1,Fintech,190
2,Healthcare,188
3,Other:,176
4,"Internet, Telecomm or Information Services",138
5,Banking/Financial Services,88
6,Government,78
7,Media & Advertising Services,75
8,Retail and Consumer Services,65
9,"Transportation, or Supply Chain",63


Висновок
Серед високооплачуваних віддалених працівників (респонденти з компенсацією вище 75-го перцентиля та форматом роботи Remote) найпоширенішими індустріями є:
Software Development — 1186, 
Fintech — 190, 
Healthcare — 188, 
Other — 176, 
Internet, Telecom or Information Services — 138.
Це свідчить про те, що найбільші можливості для високооплачуваної віддаленої роботи зосереджені в ІТ-та фінансово-технологічних галузях, а також у суміжних цифрових секторах.